internal/guest/storage/ext4/format.go

```diff
@@ -0,0 +1,186 @@
+	fin, err := os.OpenFile(source, os.O_RDONLY, 0)
+	if err != nil {
+		return errors.Wrap(err, "failed to open source file")
+	}
+	defer fin.Close()
+	fout, err := os.OpenFile(destination, os.O_WRONLY, 0)
+	if err != nil {
+		return errors.Wrap(err, "failed to open destination file")
+	}
+	defer fout.Close()
+	finInfo, err := fin.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat source file")
+	}
+	finSize := finInfo.Size()
+	var offset int64 = 0
+	for {
+		// Exit when the end of the file is reached
+		if offset >= finSize {
+			break
+		}
+		// Calculate bounds of the next data chunk
+		chunkStart, err := fin.Seek(offset, SEEK_DATA)
+		if (err != nil) || (chunkStart == -1) {
+			// No more chunks left
+			break
+		}
+		chunkEnd, err := fin.Seek(chunkStart, SEEK_HOLE)
+		if (err != nil) || (chunkEnd == -1) {
+			break
+		}
+		chunkSize := chunkEnd - chunkStart
+		offset = chunkEnd
+		// Read contents of this data chunk
+		//nolint:staticcheck //TODO: SA1019: os.SEEK_SET has been deprecated since Go 1.7: Use io.SeekStart, io.SeekCurrent, and io.SeekEnd.
+		_, err = fin.Seek(chunkStart, os.SEEK_SET)
+		if err != nil {
+			return errors.Wrap(err, "failed to seek set in source file")
+		}
+		chunkData := make([]byte, chunkSize)
+		count, err := fin.Read(chunkData)
+		if err != nil {
+			return errors.Wrap(err, "failed to read source file")
+		}
+		if int64(count) != chunkSize {
+			return errors.Wrap(err, "not enough data read from source file")
+		}
+		// Write data to destination file
+		//nolint:staticcheck //TODO: SA1019: os.SEEK_SET has been deprecated since Go 1.7: Use io.SeekStart, io.SeekCurrent, and io.SeekEnd.
+		_, err = fout.Seek(chunkStart, os.SEEK_SET)
+		if err != nil {
+			return errors.Wrap(err, "failed to seek destination file")
+		}
+		_, err = fout.Write(chunkData)
+		if err != nil {
+			return errors.Wrap(err, "failed to write destination file")
+		}
+	}
+	return nil
+}
+
+// mkfsExt4Command runs mkfs.ext4 with the provided arguments
+func mkfsExt4Command(args []string) error {
+	cmd := exec.Command("mkfs.ext4", args...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "failed to execute mkfs.ext4: %s", string(output))
+	}
+	return nil
+}
+func FormatDiskExt4(ctx context.Context, source string) error {
+	tempDir, err := _osMkdirTemp("", "ext4-dir")
+	if err != nil {
+		return errors.Wrapf(err, "failed to create temporary folder: %s", source)
+	}
+	defer func() {
+		if err := _osRemoveAll(tempDir); err != nil {
+			log.G(ctx).WithError(err).Debugf("failed to delete temporary folder: %s", tempDir)
+		}
+	}()
+	deviceSize, err := _getBlockDeviceSize(ctx, source)
+	if err != nil {
+		return fmt.Errorf("error getting size of: %s: %w", source, err)
+	}
+	if deviceSize == 0 {
+		return fmt.Errorf("invalid size obtained for: %s", source)
+	}
+	tempExt4File := filepath.Join(tempDir, "ext4.img")
+	if err = _createSparseEmptyFile(ctx, tempExt4File, deviceSize); err != nil {
```

Comment:
The requirement to format with a sparse file that is then copied to the final block device came from using dm-integrity (see the comment on `EncryptDevice` in #1090). Given we aren't using that here, I think we can avoid most of this and just call `mkfs.ext4` on the block device directly.

Reply:
I think it would be useful to experiment with what happens if the disk is already formatted or already has information on it to see if wiping the disk is something that we think we should do, especially since mkfs.ext4 appears to try to read blocks before writing to them. In which case, based on the comment in that PR, I think the sparse file option is the better option.

Reply:
I don't understand. It doesn't seem like we need to check this in to experiment with other formatting techniques in the future. Why would we use the workaround for a very specific scenario that we don't care about here (ext4 formatting a device that uses dm-integrity and dm-crypt), when we could instead just call mkfs.ext4 and check in much shorter and simpler code?

Is there something this code is doing that we actually need, other than invoking mkfs.ext4?

Reply:
This code specifically guarantees that the scratch device is empty when we've finished formatting it. This also guarantees that even if the host disk for the scratch has faulty information on it to begin with (since mkfs attempts to read the blocks before writing to them), we can still successfully complete the call. I'm not sure if we **need** these guarantees, but they might be nice to have. I'm also not sure how errors are surfaced for issues like the disk being malformed to begin with, which is why I was suggesting testing the code in that scenario.

Reply:
Can you clarify what you mean by "guarantees that the scratch device is empty" and "fault information"?

The current code does not zero-out the entire block device (which we'd want to avoid this anyway, since it would take a lot longer), so I'm not sure what you're referring to specifically.

It's likely mkfs.ext4 is not actually doing reads explicitly, and rather it is just doing block-misaligned writes, which require the OS to issue a read so it can write out the whole block at once. mkfs.ext4 is the standard way to format a block device, so I really don't think we need to overcomplicate how we're doing this here. If we know of specific relevant cases where this won't work, let's discuss those specifically.

Reply:
I see, I was misunderstanding the sparse filesystem copy, which I thought would result in old data or metadata of the disk being removed in its entirety. This doesn't appear to be the case entirely since we're only copying the metadata of the new filesystem to the disk, so that disk may still have old metadata or data partially around. 

Since without additional testing I do not have specific cases or scenarios where something like a misalignment or misconfiguration causes failures, I can remove the excess code to use sparse files.

Reply:
I've removed the code relating to spare files